connectors/sources/graphql.py

```diff
@@ -0,0 +1,333 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License 2.0;
+# you may not use this file except in compliance with the Elastic License 2.0.
+#
+"""GraphQL source module responsible to fetch documents based on GraphQL Query."""
+import json
+from copy import deepcopy
+from functools import cached_property
+
+import aiohttp
+from aiohttp.client_exceptions import ClientResponseError
+from graphql import parse
+
+from connectors.logger import logger
+from connectors.source import BaseDataSource, ConfigurableFieldValueError
+from connectors.utils import (
+    CancellableSleeps,
+    RetryStrategy,
+    hash_id,
+    iso_utc,
+    retryable,
+)
+
+RETRIES = 3
+RETRY_INTERVAL = 2
+
+PING_QUERY = """
+{
+  __schema {
+    queryType {
+      name
+    }
+  }
+}
+"""
+
+
+class UnauthorizedException(Exception):
+    pass
+
+
+class GraphQLClient:
+    def __init__(self, configuration):
+        self._sleeps = CancellableSleeps()
+        self.configuration = configuration
+        self._logger = logger
+        self.graphql_query = self.configuration["graphql_query"]
+        self.url = self.configuration["http_endpoint"]
+        self.graphql_objects = self.configuration["object_list"]
+        self.http_method = self.configuration["http_method"]
+        self.variables = {}
+        self.headers = {}
+
+    def set_logger(self, logger_):
+        self._logger = logger_
+
+    @cached_property
+    def _get_session(self):
+        timeout = aiohttp.ClientTimeout(total=300)
+        if self.configuration["authentication_method"] == "bearer":
+            self.headers.update(
+                {
+                    "Authorization": f"Bearer {self.configuration['token']}",
+                }
+            )
+            return aiohttp.ClientSession(
+                headers=self.headers,
+                timeout=timeout,
+                raise_for_status=True,
+            )
+        elif self.configuration["authentication_method"] == "basic":
+            basic_auth = aiohttp.BasicAuth(
+                login=self.configuration["username"],
+                password=self.configuration["password"],
+            )
+            return aiohttp.ClientSession(
+                auth=basic_auth,
+                headers=self.headers,
+                timeout=timeout,
+                raise_for_status=True,
+            )
+        else:
+            return aiohttp.ClientSession(
+                headers=self.headers,
+                timeout=timeout,
+                raise_for_status=True,
+            )
+
+    def boil_down_data(self, key, data):
```

Comment:
This function name is a bit vague, can we use sth like `extract_graphql_data_items` or something similar. Can we add some comments about the logic and why is it needed?